plugins/inventory/now.py

```diff
@@ -0,0 +1,443 @@
+                " is not present in table {1}.".format(
+                    inventory_hostname_source, self.get_option("table")
+                )
+            )
+
+        inventory_hostname = record[inventory_hostname_source]
+        if inventory_hostname:
+            host = self.inventory.add_host(inventory_hostname)
+            # TODO: what if the value is empty, do we handle this?
+            self.inventory.set_variable(
+                host, "ansible_host", record[ansible_host_source]
+            )
+            return host
+
+        self.display.warning(
+            "Skipping host {0} due to empty {1}".format(
+                record["sys_id"], inventory_hostname_source
+            )
+        )
+        return None
+
+    def set_hostvars(self, host, record):
+        for k, v in record.items():
+            if k in self.get_option("columns"):
+                # Set all the columns that user desired as hostvars.
+                self.inventory.set_variable(host, k, v)
+
+    def fill_auto_groups(self, table_client, table, group_by):
+        records = table_client.list_records(table, query=self._query(group_by))
+
+        for record in records:
+            host = self.add_host(record)
+            if host:
+                for category in group_by.keys():
+                    group_name = to_safe_group_name(record[category])
+                    # Only truthy names are allowed
+                    # Column values can be set to "None", and we don't want those
+                    if group_name:
+                        self.inventory.add_group(group_name)
+                        self.inventory.add_host(host, group=group_name)
+                    else:
+                        self.display.warning(
+                            "Encountered invalid group name '{1}' for host {0}. Group will not be created.".format(
+                                group_name, host
+                            )
+                        )
+
+                self.set_hostvars(host, record)
+
+    def fill_desired_groups(self, table_client, table, groups):
+        for group_name, group_conditions in groups.items():
+            self.inventory.add_group(group_name)
+
+            records = table_client.list_records(
+                table, query=self._query(group_conditions)
+            )
+            for r in records:
+                host = self.add_host(r)
+                self.inventory.add_host(host, group=group_name)
+                self.set_hostvars(host, r)
+
+    def _merge_instance_config(self, instance_config, instance_env):
+        # Pulls the values from the environment, and if necessary, overrides
+        # with configuration provided in the inventory source.
+        instance = instance_env.copy()
+        given_keys = instance_config.keys()
+        to_override = set(instance.keys()).intersection(given_keys)
+        for option in to_override:
+            instance[option] = instance_config[option]
+        return instance
```

Comment:
Maybe I am missing some context here, but if I understand the code correctly, this is just a `return dict(instance_env, **instance_config)` that skips "unknown" keys in the `instance_config` dict. In order to keep things a bit more straightforward, I would assume that the configuration file loader will not return non-declared (sub)options and that we can ignore the possibility of extra keys.

Of course, ignore my comment if my assumption about the configuration loader is off and we can get the unknown keys in the instance config.

Reply:
Yeah, the configuration file loader does return undeclared sub-options, and if we do not sanitize user input ourselves, all the extra sub-options are passed to the HTTP `Client`, which makes its `__init__` method unhappy.